tlvf: add yamls for ap_autoconfig flow

LiorAmram opened the pull request:
-Added yamls:
    eWscAttribute
    eWscValues16
    eWscValues8
    tlvWscM1
    tlvWscM2

   *Based on Wifi Alliance MultiAp specifications:
    https://groups.wi-fi.org/apps/org/workgroup/gmdocs/download.php/91680/Multi-AP_Technical_Specification-180420c.pdf

-Fixed existing yamls:
    tlvApRadioBasicCapabilities - change structure to support nested lists
    tlvApRadioIdentifier - change incorrect struct type used for mac address

-Added new and fixed yamls to tlvf_conf.yaml

Signed-off-by: Lior Amram <lior.amram@intel.com>

tomereli commented on tlvf/AutoGenerated/include/tlvf/ieee_1905_1/eWscAttributes.h:
```diff
@@ -0,0 +1,60 @@
+///////////////////////////////////////
```
Please move the autogenerated files changes to a separate commit

tomereli reviewed (changes requested):
- Please move the autogenerated files changes to a separate commit
- Please add more information in the commit message for every added .yaml file
- Please add a comment in the commit message explaining where these definitions came from (the MultiAP specification - but please use the full name or even the download link - https://groups.wi-fi.org/apps/org/workgroup/gmdocs/download.php/91680/Multi-AP_Technical_Specification-180420c.pdf)

LiorAmram committed:
tlvf: add yamls for ap_autoconfig flow

-Added yamls:
    eWscAttribute
    eWscValues16
    eWscValues8
    tlvWscM1
    tlvWscM2

   *Based on Wifi Alliance MultiAp specifications:
    https://groups.wi-fi.org/apps/org/workgroup/gmdocs/download.php/91680/Multi-AP_Technical_Specification-180420c.pdf

-Fixed existing yamls:
    tlvApRadioBasicCapabilities - change structure to support nested lists
    tlvApRadioIdentifier - change incorrect struct type used for mac address

-Added new and fixed yamls to tlvf_conf.yaml

Signed-off-by: Lior Amram <lior.amram@intel.com>

LiorAmram committed:
tlvf: add autogenerated files for autoconfig flow

-Added tlvf autogenerated .h and .cpp files for autoconfig flow

Signed-off-by: Lior Amram <lior.amram@intel.com>

arnout commented on tlvf/yaml/tlvf/ieee_1905_1/tlvWscM1.yaml:
```diff
@@ -0,0 +1,251 @@
+    _value_const: ATTR_MANUFACTURER
+  manufacturer_length:
+    _type: uint16_t
+    _value_const: WSC_MAX_NAME_LENGTH
```
This can't be right... The string `"foo\0\0\0"` is distinct from the string `"foo"`, as far as I understand, so the length should definitely not always be the maximum length.

AFAIU tlvf should support this kind of construct.

arnout commented on tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml:
```diff
@@ -0,0 +1,282 @@
+    _type: uint8_t
+    _length: [ WSC_VENDOR_EXTENSIONS_LENGTH ]
+
+  #encrypted data
```
The encrypted data should be wrapped in a KEY_WRAP_AUTH attribute. And in hostapd/wpa_supplicant, it is also wrapped in an inner CRED attribute, though the BBF implementation of 1905.1 doesn't do the latter.

LiorAmram replied:
Hi, the KEY_WRAP_AUTH should be before the encrypted data or both before and after?
Also, can you attach a reference? because I didn't see KEY_WRAP_AUTH in M2 in the WSC specs, only 'Authenticator' attribute.

arnout replied:
I was mistaken, it's wrapped in an ENCRYPTED_SETTINGS attribute, and the last (nested) attribute is KEY_WRAP_AUTH.

So I have four sources for the WSC stuff (and all of them are inconsistent :-)

1. The Multi-AP specification. It doesn't say much about how M1 and M2 have to be constructed, so not much help from there.
2. The WSC specification. This is a huge document and the information is spread all over hte place, so difficult to draw solid conclusions from it. For some reason, the M2 description (v2.0.5, section 8.3.2, p.75) doesn't mention encrypted settings, but later on tables 20 and 21 in section 8.3.9 (M8) do specify that they apply to M2 as well. The "Encrypted Settings" part in section 12 explains how this attribute is constructed, though it is a bit vague as to whether the length is just for the IV or also includes the following (encrypted) attributes. Back to Table 21, it specifies "Credential" as a required attribute, and in section 12, the Credential attribute is specified to contain SSID etc. But now I read the spec again, I see that for the M2 used in AP-Autoconfig, we have to use Table 20 instead of Table 21, and that one doesn't have "Credential", it just has SSID etc. directly there. The description of the Encrypted Settings attribute also specifies that it must end with a Key Wrap Authenticator attribute.
3. The hostapd implementation. Reading the source code, hostapd always assumes that the settings are in the Encrypted Settings attribute, and that the SSID etc. are wrapped into an inner Credential attribute. So it always follows Table 21, AFAICS.
4. The BBF meshComms implementation. This one also has the Encrypted Settings attribute, but doesn't nest into a Credential attribute.
 
Based on all of the above, my conclusion is that we need an Encrypted Settings attribute that contains (cfr. Table 20) type, length (of the entire thing), SSID, AUTH_TYPE, ENCR_TYPE, NETWORK_KEY, BSSID, KEY_WRAP_AUTH. This is exactly how meshComms constructs it. Since hostapd is never involved in this specifc M2, it's not a problem that hostapd does it differently.

LiorAmram replied:
Thank you very much!
The specs are indeed a bit confusing, but you summed it up very well.
I uploaded a PR for the refactoring (in the unified repo) and added this change in a fixup commit.

arnout commented on tlvf/yaml/tlvf/wfa_map/tlvApRadioBasicCapabilities.yaml:
```diff
@@ -9,26 +9,22 @@ tlvApRadioBasicCapabilities:
-sChannel:
-  _type: struct
-  channel: uint8_t
+    _type: uint8_t
```
I don't understand this change. I guess that means it should be in a separate commit with a proper explanation. And the rest of the changes in this file as well, I think.

LiorAmram commented on tlvf/yaml/tlvf/ieee_1905_1/eWscAttributes.yaml:
```diff
@@ -0,0 +1,39 @@
+  ATTR_NETWORK_KEY: 0x1027
+  ATTR_KEY_WRAP_AUTH: 0x101e
+  ATTR_ENCR_SETTINGS: 0x1018
+  ATTR_AUTHENTICATOR: 0x1005
```
I will, but Tomer and I didn't fully understand why it's needed

arnout replied:
It's just whitespace convention, just like the convention not to use \r\n but only \n as newlines.

The only real effect is that if you add a line at the end of the file and there was no newline, you'll get two modified lines in the diff instead of only one.

Also, some editors always add a newline at the end of the file, so if you open a file in such an editor, the last line will be changed even if you have no other modifications.

Finally, when you count the number of lines, some tools will not count that final line without newline as a line, but other tools will. I don't think this is a big deal :-)

LiorAmram replied:
Thank you for the explanation :)

LiorAmram commented on tlvf/yaml/tlvf/ieee_1905_1/eWscValues16.yaml:
```diff
@@ -0,0 +1,37 @@
+  WSC_DEV_NETWORK_INFRA_ROUTER: 2                                                                                   
+  WSC_DEV_NETWORK_INFRA_SWITCH: 3                                                                                   
+  WSC_DEV_NETWORK_INFRA_GATEWAY: 4                                                                                    
+  WSC_DEV_NETWORK_INFRA_BRIDGE: 5                                                                                   
```
I will refactor in the fixup commit.

LiorAmram commented on tlvf/yaml/tlvf/ieee_1905_1/tlvWscM1.yaml:
```diff
@@ -0,0 +1,251 @@
+    _value_const: 1
+  version:
+    _type: eWscValues8
+    _value_const: WSC_VERSION
```
I will refactor in the fixup commit.

arnout commented:
This PR has moved to https://github.com/prplfoundation/prplMesh/pull/9.

All comments have either been fixed there, or have been repeated.
